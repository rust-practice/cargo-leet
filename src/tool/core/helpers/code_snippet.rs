use super::problem_code::ProblemCode;
use crate::tool::config::Config;
use anyhow::{bail, Context};
use log::info;
use regex::Regex;

type CodeSnippets = Vec<CodeSnippet>;

#[derive(serde::Deserialize)]
struct CodeSnippetResponse {
    data: Data,
}
impl CodeSnippetResponse {
    fn into_rust_problem_code(self) -> anyhow::Result<ProblemCode> {
        let code_snippet = self.data.question.code_snippets;
        let Some(mut result) = code_snippet
            .into_iter()
            .find_map(|cs| (cs.lang == "Rust").then_some(cs.code))
        else {
            bail!("Rust not supported for this problem")
        };

        // Add todo!() placeholders in function bodies
        let re = Regex::new(r"\{\s*\}")?;
        result = re
            .replace_all(&result, r#"{ todo!("Fill in body") }"#)
            .to_string();

        result.try_into()
    }
}

#[derive(serde::Deserialize)]
struct Data {
    question: Question,
}

#[derive(serde::Deserialize)]
struct Question {
    #[serde(rename = "codeSnippet")] // TODO cw: Fix error created (missing an s)
    code_snippets: CodeSnippets,
}

#[derive(serde::Deserialize)]
struct CodeSnippet {
    lang: String,
    code: String,
}

pub(crate) fn get_code_snippet_for_problem(title_slug: &str) -> anyhow::Result<ProblemCode> {
    let request_code_snippet = request_code_snippet(title_slug)?;

    // Deserialize from Response to Struct
    let code_snippets_res = request_code_snippet
        .into_json::<CodeSnippetResponse>()
        .context("Failed to convert response from json to codes_snippet")?;

    code_snippets_res.into_rust_problem_code()
}

fn request_code_snippet(title_slug: &str) -> Result<ureq::Response, anyhow::Error> {
    info!("Going to send request for code for problem with title: {title_slug}");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": "query questionEditorData($titleSlug: String!) {
                    question(titleSlug: $titleSlug) {
                        codeSnippets {
                            lang
                            code
                        }
                    }
                }",
            "variables":{"titleSlug": title_slug},
            "operationName":"questionEditorData"
        }))
        .context("Get request for code_snippet failed")
}

#[cfg(test)]
mod tests {
    // use super::*;

    use std::io::Write as _;

    use crate::tool::core::helpers::code_snippet::request_code_snippet;

    #[test]
    fn test_name() {
        let slug = "two-sum";
        let response = request_code_snippet(slug).unwrap();
        let s = response.into_string().unwrap();
        let mut file = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .open("output.json")
            .unwrap();
        file.write_all(s.as_bytes()).unwrap();
    }
}
