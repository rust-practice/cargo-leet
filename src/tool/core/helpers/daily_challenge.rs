use anyhow::Context;
use log::info;

use crate::tool::config::Config;

use super::{get_response, local_store::path_local_store_daily_challenge};

#[derive(serde::Deserialize, Debug)]
struct DailyChallengeResponse {
    data: Data,
}
impl DailyChallengeResponse {
    fn into_title_slug(self) -> String {
        self.data
            .active_daily_coding_challenge_question
            .question
            .title_slug
    }
}

#[derive(serde::Deserialize, Debug)]
struct Data {
    #[serde(rename = "activeDailyCodingChallengeQuestion")]
    active_daily_coding_challenge_question: ActiveDailyCodingChallengeQuestion,
}

#[derive(serde::Deserialize, Debug)]
struct ActiveDailyCodingChallengeQuestion {
    question: Question,
}

#[derive(serde::Deserialize, Debug)]
struct Question {
    #[serde(rename = "titleSlug")]
    title_slug: String,
}

pub(crate) fn get_daily_challenge_slug() -> anyhow::Result<String> {
    info!("Attempting to get daily challenge");
    Ok(get_daily_challenge_response()?.into_title_slug())
}

fn get_daily_challenge_response() -> anyhow::Result<DailyChallengeResponse> {
    get_response(
        "unused_just_to_match_sig",
        local_store_request_daily_challenge,
        external_request_daily_challenge,
    )
}

fn local_store_request_daily_challenge(_needed_to_match_signature: &str) -> anyhow::Result<String> {
    let path = path_local_store_daily_challenge();
    std::fs::read_to_string(&path).with_context(|| format!("failed to read string from {path:?}"))
}

fn external_request_daily_challenge(_needed_to_match_signature: &str) -> anyhow::Result<String> {
    info!("[External] Going to send request of daily challenge");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": "query questionOfToday {
                activeDailyCodingChallengeQuestion {
                    question {
                        titleSlug
                    }
                }
            }",
            "variables":{},
            "operationName":"questionOfToday"
        }))
        .context("get request for daily challenge failed")?
        .into_string()
        .context("failed to convert response into String")
}

#[cfg(test)]
mod tests;
