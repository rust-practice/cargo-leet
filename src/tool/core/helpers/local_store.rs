//! Facilitates testing by providing local copies of responses from leetcode

use std::path::PathBuf;

const NAME_TEST_FOLDER: &str = "tests";
const NAME_LOCAL_STORE: &str = "local_store";
const NAME_LOCAL_STORE_CODE_SNIPPET: &str = "code_snippet";

pub(crate) fn path_local_store_code_snippet() -> PathBuf {
    PathBuf::from(NAME_TEST_FOLDER)
        .join(NAME_LOCAL_STORE)
        .join(NAME_LOCAL_STORE_CODE_SNIPPET)
}

#[cfg(test)]
pub(crate) mod tests {
    use rstest::fixture;

    pub(crate) type SlugList = &'static [&'static str];

    // Don't want to send them too fast and not at the same time for all tests so random in range
    pub(crate) const SECONDS_DELAY_BETWEEN_REQUESTS_MIN: u8 = 5;
    pub(crate) const SECONDS_DELAY_BETWEEN_REQUESTS_MAX: u8 = 9;

    #[fixture]
    pub(crate) fn title_slugs() -> SlugList {
        &["two-sum", "add-two-numbers", "validate-binary-search-tree"]
    }
}
