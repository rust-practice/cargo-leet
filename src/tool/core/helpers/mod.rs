pub(crate) mod code_snippet;
pub(crate) mod daily_challenge;
pub(crate) mod local_store;
pub(crate) mod problem_code;
pub(crate) mod problem_metadata;
pub(crate) mod write_to_disk;
