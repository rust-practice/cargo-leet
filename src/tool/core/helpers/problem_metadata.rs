use crate::tool::config::Config;
use anyhow::Context;
use log::info;

use super::{get_response, local_store::path_local_store_problem_metadata};

pub(crate) mod data_structure;

pub(crate) fn get_problem_metadata(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemMetadata> {
    info!("Attempting to get problem metadata");
    get_problem_metadata_response(title_slug)?.into_problem_metadata()
}

fn get_problem_metadata_response(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemMetaDataResponse> {
    get_response(
        title_slug,
        local_store_request_problem_metadata,
        external_request_problem_metadata,
    )
}

fn local_store_request_problem_metadata(title_slug: &str) -> anyhow::Result<String> {
    let path = path_local_store_problem_metadata(title_slug);
    std::fs::read_to_string(&path).with_context(|| format!("failed to read string from {path:?}"))
}

fn external_request_problem_metadata(title_slug: &str) -> anyhow::Result<String> {
    info!("[External] Going to send request for problem meta data for problem with title: {title_slug}");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": r"query consolePanelConfig($titleSlug: String!) {
            question(titleSlug: $titleSlug) {
                questionFrontendId
                questionTitle
                exampleTestcaseList
            }
        }",
            "variables":{"titleSlug": title_slug},
            "operationName":"consolePanelConfig"
        }))
        .context("failed to get request for code_snippet failed")?
        .into_string()
        .context("failed to convert response into String")
}

#[cfg(test)]
mod tests;
