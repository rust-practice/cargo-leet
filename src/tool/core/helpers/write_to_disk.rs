use anyhow::Context;
use log::{error, info};
use std::{
    env,
    fs::{remove_file, OpenOptions},
    io::Write,
    path::PathBuf,
    process::Command,
};

fn update_lib(module_name: &str) -> anyhow::Result<()> {
    info!("Adding {module_name} to libs.rs");
    let lib_path = PathBuf::from("src/lib.rs");
    let mut lib = OpenOptions::new()
        .append(true)
        .open(&lib_path)
        .with_context(|| {
            format!(
                "Failed to open {:?}",
                env::current_dir()
                    .expect("Unable to resolve current directory")
                    .join(lib_path)
            )
        })?;
    let _ = lib.write(format!("pub mod {module_name};").as_bytes())?;
    Ok(())
}

pub(crate) fn write_file(module_name: &str, module_code: &str) -> anyhow::Result<()> {
    info!("Writing code to disk for module {module_name}");
    let path = PathBuf::from(format!("src/{module_name}.rs"));
    let mut file = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&path)
        .with_context(|| format!("Failed to create '{}'", path.display()))?;
    file.write_all(module_code.as_bytes())
        .with_context(|| format!("Failed writing to '{}'", path.display()))?;
    let lib_update_status = update_lib(module_name);
    if lib_update_status.is_err() {
        error!("Failed to update lib.rs: Performing cleanup of partially completed command");
        // clean up
        remove_file(&path).with_context(|| {
            format!(
                "Failed to remove '{}' during cleanup after failing to update lib.rs",
                path.display()
            )
        })?;
        lib_update_status.context(
            "Failed to update lib.rs. Does the file exists? Is it able to be written to?",
        )?;
    }

    info!("Going to run rustfmt on files");
    Command::new("cargo")
        .arg("fmt")
        .arg("--all")
        .output()
        .context("Error running rustfmt")?;
    Ok(())
}
