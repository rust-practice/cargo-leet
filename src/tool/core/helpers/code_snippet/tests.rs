use std::io::Write as _;

use anyhow::Context;
use rstest::rstest;

use crate::tool::core::helpers::{
    code_snippet::{external_request_code_snippet, get_code_snippets_response},
    local_store::tests::{get_rnd_request_delay, title_slugs, SlugList},
};

#[rstest]
#[ignore = "Only use for downloading responses"]
fn download_response_from_leetcode(title_slugs: SlugList) {
    for title_slug in title_slugs {
        let sleep_delay = std::time::Duration::from_millis(get_rnd_request_delay());
        println!(
            "Going to sleep for {} milliseconds before requesting and trying to save {title_slug}",
            sleep_delay.as_millis()
        );
        std::thread::sleep(sleep_delay); // Sleep to not go too hard on leetcode API
        let response_string = external_request_code_snippet(title_slug).unwrap();
        let path = super::path_local_store_code_snippet(title_slug);
        let mut file = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(&path)
            .with_context(|| format!("failed to save response to {path:?}"))
            .unwrap();
        file.write_all(response_string.as_bytes()).unwrap();
        println!("Save of {title_slug:?} completed\n");
    }
    println!("Successfully saved all responses");
}

#[rstest]
fn conversion_from_leetcode_response(title_slugs: SlugList) {
    for title_slug in title_slugs {
        insta::assert_debug_snapshot!(get_code_snippets_response(title_slug).unwrap());
    }
}
