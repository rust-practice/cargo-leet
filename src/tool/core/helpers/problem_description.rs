use crate::tool::config::Config;
use anyhow::Context;
use log::info;

use super::{get_response, local_store::path_local_store_problem_description};

pub(crate) mod data_structure;

pub(crate) fn get_problem_description(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemDescription> {
    info!("Attempting to get problem description");
    get_problem_description_response(title_slug)?.try_into()
}

fn get_problem_description_response(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemDescriptionResponse> {
    get_response(
        title_slug,
        local_store_request_problem_description,
        external_request_problem_description,
    )
}

fn local_store_request_problem_description(title_slug: &str) -> anyhow::Result<String> {
    let path = path_local_store_problem_description(title_slug);
    std::fs::read_to_string(&path).with_context(|| format!("failed to read string from {path:?}"))
}

fn external_request_problem_description(title_slug: &str) -> anyhow::Result<String> {
    info!("[External] Going to send request for problem description for problem with title: {title_slug}");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": r"query questionContent($titleSlug: String!) {
            question(titleSlug: $titleSlug) {
                content
            }
        }",
            "variables":{"titleSlug": title_slug},
            "operationName":"questionContent"
        }))
        .context("failed to get request for description failed")?
        .into_string()
        .context("failed to convert response into String")
}

#[cfg(test)]
mod tests {
    use std::io::Write as _;

    use anyhow::Context;
    use rstest::rstest;

    use crate::tool::core::helpers::{
        local_store::{
            path_local_store_problem_description,
            tests::{get_rnd_request_delay, insta_settings, title_slugs, SlugList},
        },
        problem_description::{external_request_problem_description, get_problem_description},
    };

    #[rstest]
    #[ignore = "Only use for downloading responses"]
    fn download_response_from_leetcode(title_slugs: SlugList) {
        for title_slug in title_slugs {
            let sleep_delay = std::time::Duration::from_millis(get_rnd_request_delay());
            println!(
            "Going to sleep for {} milliseconds before requesting and trying to save {title_slug}",
            sleep_delay.as_millis()
        );
            std::thread::sleep(sleep_delay); // Sleep to not go too hard on leetcode API
            let response_string = external_request_problem_description(title_slug).unwrap();
            let path = path_local_store_problem_description(title_slug);
            let mut file = std::fs::OpenOptions::new()
                .write(true)
                .create(true)
                .truncate(true)
                .open(&path)
                .with_context(|| format!("failed to save response to {path:?}"))
                .unwrap();
            file.write_all(response_string.as_bytes()).unwrap();
            println!("Save of {title_slug:?} completed\n");
        }
        println!("Successfully saved all responses");
    }

    #[rstest]
    fn conversion_from_leetcode_response(title_slugs: SlugList, insta_settings: insta::Settings) {
        for title_slug in title_slugs {
            insta_settings.bind(|| {
                insta::assert_debug_snapshot!(get_problem_description(title_slug).unwrap());
            });
        }
    }

    #[rstest]
    fn extract_solutions_from_description(title_slugs: SlugList, insta_settings: insta::Settings) {
        for title_slug in title_slugs {
            insta_settings.bind(|| {
                let problem_description = get_problem_description(title_slug).unwrap();
                insta::assert_debug_snapshot!(problem_description.get_solutions());
            });
        }
    }
}
