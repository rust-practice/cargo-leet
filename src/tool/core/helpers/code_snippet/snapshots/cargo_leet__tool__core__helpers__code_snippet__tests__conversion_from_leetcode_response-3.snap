---
source: src/tool/core/helpers/code_snippet/tests.rs
expression: get_code_snippets_response(title_slug).unwrap()
---
CodeSnippetResponse {
    data: Data {
        question: Question {
            code_snippets: [
                CodeSnippet {
                    lang: "C++",
                    code: "/**\n * Definition for a binary tree node.\n * struct TreeNode {\n *     int val;\n *     TreeNode *left;\n *     TreeNode *right;\n *     TreeNode() : val(0), left(nullptr), right(nullptr) {}\n *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}\n *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}\n * };\n */\nclass Solution {\npublic:\n    bool isValidBST(TreeNode* root) {\n        \n    }\n};",
                },
                CodeSnippet {
                    lang: "Java",
                    code: "/**\n * Definition for a binary tree node.\n * public class TreeNode {\n *     int val;\n *     TreeNode left;\n *     TreeNode right;\n *     TreeNode() {}\n *     TreeNode(int val) { this.val = val; }\n *     TreeNode(int val, TreeNode left, TreeNode right) {\n *         this.val = val;\n *         this.left = left;\n *         this.right = right;\n *     }\n * }\n */\nclass Solution {\n    public boolean isValidBST(TreeNode root) {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "Python",
                    code: "# Definition for a binary tree node.\n# class TreeNode(object):\n#     def __init__(self, val=0, left=None, right=None):\n#         self.val = val\n#         self.left = left\n#         self.right = right\nclass Solution(object):\n    def isValidBST(self, root):\n        \"\"\"\n        :type root: TreeNode\n        :rtype: bool\n        \"\"\"\n        ",
                },
                CodeSnippet {
                    lang: "Python3",
                    code: "# Definition for a binary tree node.\n# class TreeNode:\n#     def __init__(self, val=0, left=None, right=None):\n#         self.val = val\n#         self.left = left\n#         self.right = right\nclass Solution:\n    def isValidBST(self, root: Optional[TreeNode]) -> bool:\n        ",
                },
                CodeSnippet {
                    lang: "C",
                    code: "/**\n * Definition for a binary tree node.\n * struct TreeNode {\n *     int val;\n *     struct TreeNode *left;\n *     struct TreeNode *right;\n * };\n */\nbool isValidBST(struct TreeNode* root) {\n    \n}",
                },
                CodeSnippet {
                    lang: "C#",
                    code: "/**\n * Definition for a binary tree node.\n * public class TreeNode {\n *     public int val;\n *     public TreeNode left;\n *     public TreeNode right;\n *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {\n *         this.val = val;\n *         this.left = left;\n *         this.right = right;\n *     }\n * }\n */\npublic class Solution {\n    public bool IsValidBST(TreeNode root) {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "JavaScript",
                    code: "/**\n * Definition for a binary tree node.\n * function TreeNode(val, left, right) {\n *     this.val = (val===undefined ? 0 : val)\n *     this.left = (left===undefined ? null : left)\n *     this.right = (right===undefined ? null : right)\n * }\n */\n/**\n * @param {TreeNode} root\n * @return {boolean}\n */\nvar isValidBST = function(root) {\n    \n};",
                },
                CodeSnippet {
                    lang: "TypeScript",
                    code: "/**\n * Definition for a binary tree node.\n * class TreeNode {\n *     val: number\n *     left: TreeNode | null\n *     right: TreeNode | null\n *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {\n *         this.val = (val===undefined ? 0 : val)\n *         this.left = (left===undefined ? null : left)\n *         this.right = (right===undefined ? null : right)\n *     }\n * }\n */\n\nfunction isValidBST(root: TreeNode | null): boolean {\n    \n};",
                },
                CodeSnippet {
                    lang: "PHP",
                    code: "/**\n * Definition for a binary tree node.\n * class TreeNode {\n *     public $val = null;\n *     public $left = null;\n *     public $right = null;\n *     function __construct($val = 0, $left = null, $right = null) {\n *         $this->val = $val;\n *         $this->left = $left;\n *         $this->right = $right;\n *     }\n * }\n */\nclass Solution {\n\n    /**\n     * @param TreeNode $root\n     * @return Boolean\n     */\n    function isValidBST($root) {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "Swift",
                    code: "/**\n * Definition for a binary tree node.\n * public class TreeNode {\n *     public var val: Int\n *     public var left: TreeNode?\n *     public var right: TreeNode?\n *     public init() { self.val = 0; self.left = nil; self.right = nil; }\n *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }\n *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {\n *         self.val = val\n *         self.left = left\n *         self.right = right\n *     }\n * }\n */\nclass Solution {\n    func isValidBST(_ root: TreeNode?) -> Bool {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "Kotlin",
                    code: "/**\n * Example:\n * var ti = TreeNode(5)\n * var v = ti.`val`\n * Definition for a binary tree node.\n * class TreeNode(var `val`: Int) {\n *     var left: TreeNode? = null\n *     var right: TreeNode? = null\n * }\n */\nclass Solution {\n    fun isValidBST(root: TreeNode?): Boolean {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "Dart",
                    code: "/**\n * Definition for a binary tree node.\n * class TreeNode {\n *   int val;\n *   TreeNode? left;\n *   TreeNode? right;\n *   TreeNode([this.val = 0, this.left, this.right]);\n * }\n */\nclass Solution {\n  bool isValidBST(TreeNode? root) {\n    \n  }\n}",
                },
                CodeSnippet {
                    lang: "Go",
                    code: "/**\n * Definition for a binary tree node.\n * type TreeNode struct {\n *     Val int\n *     Left *TreeNode\n *     Right *TreeNode\n * }\n */\nfunc isValidBST(root *TreeNode) bool {\n    \n}",
                },
                CodeSnippet {
                    lang: "Ruby",
                    code: "# Definition for a binary tree node.\n# class TreeNode\n#     attr_accessor :val, :left, :right\n#     def initialize(val = 0, left = nil, right = nil)\n#         @val = val\n#         @left = left\n#         @right = right\n#     end\n# end\n# @param {TreeNode} root\n# @return {Boolean}\ndef is_valid_bst(root)\n    \nend",
                },
                CodeSnippet {
                    lang: "Scala",
                    code: "/**\n * Definition for a binary tree node.\n * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {\n *   var value: Int = _value\n *   var left: TreeNode = _left\n *   var right: TreeNode = _right\n * }\n */\nobject Solution {\n    def isValidBST(root: TreeNode): Boolean = {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "Rust",
                    code: "// Definition for a binary tree node.\n// #[derive(Debug, PartialEq, Eq)]\n// pub struct TreeNode {\n//   pub val: i32,\n//   pub left: Option<Rc<RefCell<TreeNode>>>,\n//   pub right: Option<Rc<RefCell<TreeNode>>>,\n// }\n// \n// impl TreeNode {\n//   #[inline]\n//   pub fn new(val: i32) -> Self {\n//     TreeNode {\n//       val,\n//       left: None,\n//       right: None\n//     }\n//   }\n// }\nuse std::rc::Rc;\nuse std::cell::RefCell;\nimpl Solution {\n    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {\n        \n    }\n}",
                },
                CodeSnippet {
                    lang: "Racket",
                    code: "; Definition for a binary tree node.\n#|\n\n; val : integer?\n; left : (or/c tree-node? #f)\n; right : (or/c tree-node? #f)\n(struct tree-node\n  (val left right) #:mutable #:transparent)\n\n; constructor\n(define (make-tree-node [val 0])\n  (tree-node val #f #f))\n\n|#\n\n(define/contract (is-valid-bst root)\n  (-> (or/c tree-node? #f) boolean?)\n  )",
                },
                CodeSnippet {
                    lang: "Erlang",
                    code: "%% Definition for a binary tree node.\n%%\n%% -record(tree_node, {val = 0 :: integer(),\n%%                     left = null  :: 'null' | #tree_node{},\n%%                     right = null :: 'null' | #tree_node{}}).\n\n-spec is_valid_bst(Root :: #tree_node{} | null) -> boolean().\nis_valid_bst(Root) ->\n  .",
                },
                CodeSnippet {
                    lang: "Elixir",
                    code: "# Definition for a binary tree node.\n#\n# defmodule TreeNode do\n#   @type t :: %__MODULE__{\n#           val: integer,\n#           left: TreeNode.t() | nil,\n#           right: TreeNode.t() | nil\n#         }\n#   defstruct val: 0, left: nil, right: nil\n# end\n\ndefmodule Solution do\n  @spec is_valid_bst(root :: TreeNode.t | nil) :: boolean\n  def is_valid_bst(root) do\n    \n  end\nend",
                },
            ],
        },
    },
}
